xarray/core/variable.py

```diff
@@ -1212,34 +1212,79 @@ def __setitem__(self, key, value):
         raise TypeError('%s values cannot be modified' % type(self).__name__)
 
     @classmethod
+    def _concat_numpy(cls, variables, positions=None):
+        """
+        Concatenates variables. Works for variables whose dtype is
+        different from numpy.object. If variables' dtype is numpy.object
+        it throws TypeError and "concat" function will use
+        concat_pandas function
+        :param variables: list of variables to concatenate
+        :return: Concatenated variables
+        """
+
+        indexes = [v._data for v in variables]
+
+        if not indexes:
+            data = []
+        else:
+            data = np.concatenate((indexes))
+
+            if positions is not None:
+                indices = nputils.inverse_permutation(
+                    np.concatenate(positions))
+                data = data.take(indices)
+
+        return data
+
+    @classmethod
+    def _concat_pandas(cls, variables, positions=None):
+        """
+        Concatenates variables. This is generic function that
+        handles all cases for which concat_numpy does not work
+        :param variables: list of variables to concatenate
+        :return: Concatenated variables
+        """
+        indexes = [v._data.array for v in variables]
+
+        if not indexes:
+            data = []
+        else:
+
+            data = indexes[0].append(indexes[1:])
+
+            if positions is not None:
```

Comment:
This block is the same for both `_concat_pandas` and `_concat_numpy` -- please move it up into the caller instead.